File: finanz-api/src/main/java/com/finanzapi/service/BudgetService.java
package com.finanzapi.service;

import com.finanzapi.model.TipoTransaccion;
import com.finanzapi.model.Transaccion;
import com.finanzapi.repository.TransaccionRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Business logic for the Finanz API.
 *
 * <p>The Spring Boot counterpart to {@code finanz-core}'s {@code BudgetService} —
 * same domain logic, now backed by a JPA {@link TransaccionRepository} instead of
 * a CSV file. The service sits between {@code BudgetController} and the repository,
 * keeping HTTP concerns out of the business layer.</p>
 *
 * <h3>Architecture note</h3>
 * <p>Both {@code registrar} and {@code registrarSimple} persist via Spring Data JPA.
 * The difference is validation: {@code registrarSimple} builds the entity in the
 * service from validated primitives, whereas {@code registrar} accepts the full
 * entity from the controller (for integration-friendly use).</p>
 */
@Service
public class BudgetService {

    private final TransaccionRepository repositorio;

    /**
     * @param repositorio the JPA repository backed by H2 ({@code transactions} table)
     */
    public BudgetService(TransaccionRepository repositorio) {
        this.repositorio = repositorio;
    }

    /**
     * Persists a full transaction, overriding its date to today.
     *
     * @param transaccion the transaction to save; its {@code fecha} will be set to now
     * @return the saved entity with auto-assigned ID and today's date
     */
    public Transaccion registrar(Transaccion transaccion) {
        transaccion.setFecha(LocalDate.now());
        return repositorio.save(transaccion);
    }

    /**
     * Builds and persists a transaction from primitive inputs.
     *
     * <p>Intended for the {@code /transacciones/simple} endpoint; the controller
     * validates inputs before calling this method.</p>
     *
     * @param descripcion a short description of the transaction
     * @param monto       the absolute amount (positive)
     * @param tipo        {@code INGRESO} (income) or {@code GASTO} (expense)
     * @return the saved transaction entity with auto-assigned ID
     */
    public Transaccion registrarSimple(String descripcion, double monto, TipoTransaccion tipo) {
        Transaccion transaccion = new Transaccion(
                descripcion,
                monto,
                tipo,
                LocalDate.now()
        );
        return repositorio.save(transaccion);
    }

    /**
     * Returns all recorded transactions.
     *
     * @return a list of all transactions; empty list if none exist
     */
    public List<Transaccion> obtenerTodas() {
        return repositorio.findAll();
    }

    /**
     * Retrieves a single transaction by its primary key.
     *
     * @param id the transaction ID
     * @return an {@link Optional} with the transaction, or empty if not found
     */
    public Optional<Transaccion> obtenerPorId(Long id) {
        return repositorio.findById(id);
    }

    /**
     * Deletes the transaction with the given ID.
     *
     * @param id the transaction ID
     * @return {@code true} if the transaction existed and was deleted;
     *         {@code false} if no matching ID was found
     */
    public boolean eliminar(Long id) {
        if (repositorio.existsById(id)) {
            repositorio.deleteById(id);
            return true;
        }
        return false;
    }

    /**
     * Calculates the all-time net balance (total income minus total expenses).
     *
     * <p>Uses {@link TransaccionRepository#findByTipo} — a Spring Data derived
     * query — to avoid loading all transactions when type filtering is needed.</p>
     *
     * @return the net balance across all recorded transactions
     */
    public double calcularSaldo() {
        double ingresos = repositorio.findByTipo(TipoTransaccion.INGRESO)
                .stream()
                .mapToDouble(Transaccion::getMonto)
                .sum();

        double gastos = repositorio.findByTipo(TipoTransaccion.GASTO)
                .stream()
                .mapToDouble(Transaccion::getMonto)
                .sum();

        return ingresos - gastos;
    }

    /**
     * Calculates the net balance for the current calendar month.
     *
     * <p>Uses {@link TransaccionRepository#findByFechaBetween} (first day of
     * current month → today) to restrict the query to this month's transactions.</p>
     *
     * @return net balance for the current month (income minus expenses)
     */
    public double calcularSaldoMesActual() {
        LocalDate inicio = LocalDate.now().withDayOfMonth(1);
        LocalDate fin = LocalDate.now();

        List<Transaccion> transacciones = repositorio.findByFechaBetween(inicio, fin);

        double ingresos = transacciones.stream()
                .filter(t -> t.getTipo() == TipoTransaccion.INGRESO)
                .mapToDouble(Transaccion::getMonto)
                .sum();

        double gastos = transacciones.stream()
                .filter(t -> t.getTipo() == TipoTransaccion.GASTO)
                .mapToDouble(Transaccion::getMonto)
                .sum();

        return ingresos - gastos;
    }

    /**
     * Builds a full monthly summary map for the current calendar month.
     *
     * <p>Response map keys:</p>
     * <ul>
     *   <li>{@code mes} — current month number (1–12)</li>
     *   <li>{@code anio} — current year</li>
     *   <li>{@code ingresos} — total income this month</li>
     *   <li>{@code gastos} — total expenses this month</li>
     *   <li>{@code saldo} — net balance ({@code ingresos − gastos})</li>
     *   <li>{@code transacciones} — list of all transactions this month</li>
     * </ul>
     *
     * @return a {@link Map} with the keys described above
     */
    public Map<String, Object> obtenerResumenMensual() {
        LocalDate inicio = LocalDate.now().withDayOfMonth(1);
        LocalDate fin = LocalDate.now();

        List<Transaccion> transacciones = repositorio.findByFechaBetween(inicio, fin);

        double ingresos = transacciones.stream()
                .filter(t -> t.getTipo() == TipoTransaccion.INGRESO)
                .mapToDouble(Transaccion::getMonto)
                .sum();

        double gastos = transacciones.stream()
                .filter(t -> t.getTipo() == TipoTransaccion.GASTO)
                .mapToDouble(Transaccion::getMonto)
                .sum();

        Map<String, Object> resumen = new HashMap<>();
        resumen.put("mes", LocalDate.now().getMonthValue());
        resumen.put("anio", LocalDate.now().getYear());
        resumen.put("ingresos", ingresos);
        resumen.put("gastos", gastos);
        resumen.put("saldo", ingresos - gastos);
        resumen.put("transacciones", transacciones);

        return resumen;
    }
}

File: finanz-app/app/src/main/java/com/finanzapp/domain/BudgetService.kt
package com.finanzapp.domain

import com.finanzapp.data.local.TransaccionRepository
import com.finanzapp.data.model.Categoria
import com.finanzapp.data.model.TipoTransaccion
import com.finanzapp.data.model.Transaccion
import java.time.LocalDate

/**
 * Business logic for the FinanzApp budget domain.
 *
 * Sits between [com.finanzapp.ui.viewmodel.BudgetViewModel] and [TransaccionRepository],
 * keeping financial calculations out of the ViewModel. All methods are `suspend` and
 * must be called from a coroutine scope.
 *
 * This service is the Kotlin/Android counterpart of `finanz-core`'s `BudgetService`
 * and `finanz-api`'s `BudgetService` — same domain logic, different delivery mechanism.
 */
class BudgetService(private val repositorio: TransaccionRepository) {

    /**
     * Records a new transaction with today's date.
     *
     * [Categoria] is always set to `"General"` / `0.0` for domain compatibility;
     * the UI does not expose category or budget-limit input.
     *
     * @param descripcion a short description of the movement
     * @param monto       the absolute amount (positive; direction is set by [tipo])
     * @param tipo        [TipoTransaccion.INGRESO] for income or [TipoTransaccion.GASTO] for expense
     * @return the auto-generated row ID assigned by Room
     */
    suspend fun registrarTransaccion(
        descripcion: String,
        monto: Double,
        tipo: TipoTransaccion
    ): Long {
        val transaccion = Transaccion(
            descripcion = descripcion,
            monto = monto,
            tipo = tipo,
            fecha = LocalDate.now(),
            categoria = Categoria(
                nombre = "General",
                presupuestoLimite = 0.0
            )
        )
        return repositorio.insertar(transaccion)
    }

    /**
     * Calculates the all-time net balance (total income minus total expenses).
     *
     * @return net balance across all recorded transactions
     */
    suspend fun calcularSaldo(): Double {
        val todas = repositorio.obtenerTodas()
        val ingresos = todas.filter { it.tipo == TipoTransaccion.INGRESO }.sumOf { it.monto }
        val gastos = todas.filter { it.tipo == TipoTransaccion.GASTO }.sumOf { it.monto }
        return ingresos - gastos
    }

    /**
     * Calculates the net balance for the current calendar month.
     *
     * @return net balance from the first day of the current month to today
     */
    suspend fun calcularSaldoMesActual(): Double {
        val inicio = LocalDate.now().withDayOfMonth(1)
        val fin = LocalDate.now()
        val transacciones = repositorio.obtenerEntreFechas(inicio, fin)
        val ingresos = transacciones.filter { it.tipo == TipoTransaccion.INGRESO }.sumOf { it.monto }
        val gastos = transacciones.filter { it.tipo == TipoTransaccion.GASTO }.sumOf { it.monto }
        return ingresos - gastos
    }

    /**
     * Builds a full [ResumenMensual] for the current calendar month.
     *
     * @return aggregated income, expenses, balance, and transaction list for this month
     */
    suspend fun obtenerResumenMensual(): ResumenMensual {
        val inicio = LocalDate.now().withDayOfMonth(1)
        val fin = LocalDate.now()
        val transacciones = repositorio.obtenerEntreFechas(inicio, fin)
        val ingresos = transacciones.filter { it.tipo == TipoTransaccion.INGRESO }.sumOf { it.monto }
        val gastos = transacciones.filter { it.tipo == TipoTransaccion.GASTO }.sumOf { it.monto }
        return ResumenMensual(
            mes = LocalDate.now().monthValue,
            anio = LocalDate.now().year,
            ingresos = ingresos,
            gastos = gastos,
            saldo = ingresos - gastos,
            transacciones = transacciones
        )
    }

    /**
     * Deletes the transaction with the given [id].
     *
     * @param id primary key of the transaction to remove
     */
    suspend fun eliminarTransaccion(id: Long) = repositorio.eliminar(id)
}

/**
 * Aggregated monthly financial figures for the current calendar month.
 *
 * @property mes          current month number (1–12)
 * @property anio         current year
 * @property ingresos     total income this month
 * @property gastos       total expenses this month
 * @property saldo        net balance (ingresos − gastos)
 * @property transacciones all transactions recorded this month
 */
data class ResumenMensual(
    val mes: Int,
    val anio: Int,
    val ingresos: Double,
    val gastos: Double,
    val saldo: Double,
    val transacciones: List<Transaccion>
)
